docs/architecture-overview.rst

```diff
@@ -0,0 +1,200 @@
+come in from outside.
+
+- All requests to URLs beginning with ``/static/`` are served from the
+  corresponding files in ``/home/zulip/prod-static/``.
+
+- Requests to ``/json/get_events``, ``/api/v1/events``, and
+  ``/sockjs`` are sent to the Tornado server running at
+  ``http://tornado``. These are requests to the real-time push system,
+  because the user's web browser sets up a long-lived TCP connection
+  with Tornado to serve as `a channel for push notifications <
+  https://en.wikipedia.org/wiki/Push_technology#Long_Polling>`__.
+
+- Requests to ``/api`` and to ``/`` (i.e., everything else) are sent
+  to the Django app via the UNIX socket
+  ``unix:/home/zulip/deployments/fastcgi-socket`` (defined in
+  ``puppet/zulip/files/nginx/zulip-include-frontend/upstreams``).
+
+
+
+Supervisor
+----------
+
+Starts processes and restarts them if they crash. The config file is
+``zulip/puppet/zulip/files/supervisor/conf.d/zulip.conf``. This is where
+Tornado and Django are set up, as well as a number of background
+processes that process event queues.
+
+
+memcached
+---------
+
+memcached is used to cache database model
+objects. ``zerver/lib/cache.py`` and ``zerver/lib/cache_helpers.py``
+manage putting things into memcached, and its configuration is in in
+``puppet/zulip/files/memcached.conf``.
+
+Redis
+-----
+
+Redis is used for a few very short-term data stores, such as in the
+basis of ``zerver/lib/rate_limiter.py``, a per-user rate limiting
+scheme `example
+<http://blog.domaintools.com/2013/04/rate-limiting-with-redis/>`__).
+
+Redis is configured in ``zulip/puppet/zulip/files/redis`` and it's a
+pretty standard configuration except for the last line, which turns
+off persistence:
+
+::
+
+     # Zulip-specific configuration: disable saving to disk.
+     save ""
+
+memcached was used first and then we added Redis specifically to
+implement rate limiting. `We'd like to switch everything over to
+Redis.<https://github.com/zulip/zulip/issues/16>`__
+
+
+RabbitMQ
+--------
+
+RabbitMQ is a queueing system. Its config files live in
+``zulip/puppet/zulip/files/rabbitmq``. Initial configuration happens
+in ``zulip/scripts/setup/configure-rabbitmq``.
+
+We use RabbitMQ for queuing expensive work (e.g. sending emails
+triggered by a message, push notifications, some analytics, etc.) that
+require reliable delivery but which we don't want to do on the main
+thread. It's also used for communication between the application
+server and the Tornado push system.
+
+Two simple wrappers around ``pika`` (the Python RabbitMQ client) are
+in ``zulip/server/lib/queue.py``. There's an asynchronous client for
+use in Tornado and a more general client for use elsewhere.
+
+``zerver/lib/event_queue.py`` has helper functions for putting events
+into one queue or another. Most of the processes started by Supervise
+are queue processors that continually pull things out of a RabbitMQ
+queue and handle them.
```

Comment:
This section should probably link to the more extensive docs here: http://zulip.readthedocs.io/en/latest/queuing.html